launchflow/gcp/gke_service.py

```diff
@@ -0,0 +1,262 @@
+        - `service_type (Literal["ClusterIP", "NodePort", "LoadBalancer"])`: The type of Kubernetes service to create. Defaults to `LoadBalancer`.
+        - `startup_probe (Optional[StartupProbe])`: The [startup probe](/reference/kubernetes-resources/service#startup-probe) for the service.
+        - `liveness_probe (Optional[LivenessProbe])`: The [liveness probe](/reference/kubernetes-resources/service#liveness-probe) for the service.
+        - `readiness_probe (Optional[ReadinessProbe])`: The [readiness probe](/reference/kubernetes-resources/service#readiness-probe) for the service.
+        - `environment_variables (Optional[Dict[str, str]])`: A dictionary of environment variables to set on the service container.
+        - `container_resources (Optional[ContainerResources])`: The [container resources](/reference/kubernetes-resources/service#container-resources) for the service. NOTE: If you are creating a [HorizontalPodAutoscaler](/reference/kubernetes-resources/hpa) you will need to provide this.
+        - `tolerations (Optional[List[Toleration]])`: A list of [tolerations](/reference/kubernetes-resources/service#toleration) for the service.
+        - `domain (Optional[str])`: The domain to use for the service. `service_type` must be `NodePort` to assign a custom domain.
+        """
+        super().__init__(
+            name=name,
+            dockerfile=dockerfile,
+            build_directory=build_directory,
+            build_ignore=build_ignore,
+        )
+
+        self._artifact_registry = ArtifactRegistryRepository(
+            f"{name}-repository", format=RegistryFormat.DOCKER
+        )
+        self.container = ServiceContainer(
+            name,
+            cluster=cluster,
+            namespace=namespace,
+            node_pool=node_pool,
+            container_port=container_port,
+            host_port=host_port,
+            startup_probe=startup_probe,
+            liveness_probe=liveness_probe,
+            readiness_probe=readiness_probe,
+            service_type=service_type,
+            container_resources=container_resources,
+            tolerations=tolerations,
+        )
+        self.cluster = cluster
+        self.namespace = namespace
+        self.environment_variables = environment_variables
+        self.domain = domain
+        self._custom_domain_mapping = None
+        self._ip_address = None
+        self._ssl_certificate = None
+        if domain:
+            if service_type != "NodePort":
+                raise ValueError(
+                    "Custom domains are only supported for ClusterIP services. Please set service_type='ClusterIP' to use a custom domain."
+                )
+            self._ip_address = GlobalIPAddress(f"{name}-ip-address")
+            self._ssl_certificate = ManagedSSLCertificate(
```

Comment:
dont need to do right now, but we'll need to figure out a way to let users customize this. Lambda has a similar issue, maybe we could brainstorm a pattern to use for these cases

Reply:
Yea I _think_ the general patter will be just to allow users to pass them in to the service instead of having the service define them. So if you started with having the service manage them but need more customization you could pull them out and pass them in instead.

Reply:
or we could add a method for setting them or something